### PR TITLE
Fix UCR report builder "Edit Report" bug

### PR DESCRIPTION
**Pull request by NoahCarnahan**
http://manage.dimagi.com/default.asp?171118
An error could occur when users edited reports that had been built with the builder than were based on a form that had changed. If the report contained columns for questions that no longer exist, the user would get a 500.

In a related bug, updating an existing report to include questions that had been added since the report was originally created would cause errors because the data source backing the report was not updated.

This PR fixes both of those issues. Can be reviewed commit-by-commit.

cc @gcapalbo

### CODE REVIEW

File: corehq/apps/userreports/reports/builder/forms.py [hunk 1]
```diff
@@ -430,7 +432,63 @@ def filter_fieldset(self):
...
+        matching_data_source = self.ds_builder.get_existing_match()
+        if matching_data_source:
+            if matching_data_source['id'] != self.existing_report.config_id:
+
+                # If no one else is using the current data source, delete it.
+                data_source = DataSourceConfiguration.get(self.existing_report.config_id)
+                if data_source.get_report_count() <= 1:
+                    try:
+                        delete_data_source_shared(self.domain, data_source._id)
+                    except Exception as e:
```
Comment (dimagimon) [inline corehq/apps/userreports/reports/builder/forms.py]: F841 local variable 'e' is assigned to but never used

File: corehq/apps/userreports/reports/builder/forms.py [hunk 2]
```diff
@@ -430,7 +432,63 @@ def filter_fieldset(self):
...
+                        import ipdb; ipdb.set_trace()
```
Comment (dimagimon) [inline corehq/apps/userreports/reports/builder/forms.py]: E702 multiple statements on one line (semicolon)

**Comment by gcapalbo** [comment]
(build stalled, restarted it)

**Comment by gcapalbo** [comment]
Looks good to me - does this need any additional testing?  If not, it's good to merge

File: corehq/apps/userreports/tests/test_report_builder.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import os
+from django.test import TestCase
+from corehq.apps.app_manager.const import APP_V2
+from corehq.apps.app_manager.models import Application, Module
+from corehq.apps.userreports.models import DataSourceConfiguration, ReportConfiguration
+from corehq.apps.userreports.reports.builder.forms import ConfigureListReportForm
+
+
+
+def read(rel_path):
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_report_builder.py]: E303 too many blank lines (3)

**Comment by NoahCarnahan** [comment]
@gcapalbo Added a test. Thanks for the suggestion.

**Comment by gcapalbo** [comment]
nice! just waiting for the build and will merge

File: corehq/apps/userreports/models.py [hunk 4]
```diff
@@ -198,6 +198,16 @@ def get_all_values(self, doc):
             for i, item in enumerate(self.get_items(doc))
         ]
 
+    def get_report_count(self):
+        """
+        Return the number of ReportConfigurations that reference this data source.
+        """
+        return DataSourceConfiguration.view(
```
Comment (czue) [inline corehq/apps/userreports/models.py]: nitpick, but this should probably be done through a `dbacessors` call

Comment (NoahCarnahan) [inline corehq/apps/userreports/models.py]: Fixed here: https://github.com/dimagi/commcare-hq/pull/7153

Comment (czue) [inline corehq/apps/userreports/models.py]: thanks!

**Pull request merged**